File: CsharpProjects/myLeetCode/Stack_Queue/20. Valid Parentheses.cs
// 20. Valid Parentheses
// https://leetcode.com/problems/valid-parentheses/

using System.Collections.Generic;

public class Solution {
    public bool IsValid(string s) {
        Stack<char> stack = new Stack<char>();

        foreach (char c in s) {
            if (c == '(' || c == '{' || c == '[') {
                stack.Push(c);
            } else {
                if (stack.Count == 0) return false;
                char top = stack.Pop();
                if (c == ')' && top != '(') return false;
                if (c == '}' && top != '{') return false;
                if (c == ']' && top != '[') return false;
            }
        }

        return stack.Count == 0;
    }
}
public class Program {
    public static void Main() {
        Solution solution = new Solution();

        string input = "{[]}";
        bool result = solution.IsValid(input);

        Console.WriteLine($"Input: '{input}'");
        Console.WriteLine($"Is valid: {result}");
    }
}

File: C++/myLeetCode/Stack_Queue/20. Valid Parentheses.cpp
// 20. Valid Parentheses
// https://leetcode.com/problems/valid-parentheses/

#include <iostream>
#include <stack>

bool isValid(std::string s) {
    std::stack<char> stk;
    for (char c : s) {
        if (c == '(' || c == '{' || c == '[') {
            stk.push(c);
        } else {
            if (stk.empty()) return false;
            if (c == ')' && stk.top() != '(') return false;
            if (c == '}' && stk.top() != '{') return false;
            if (c == ']' && stk.top() != '[') return false;
            stk.pop();
        }
    }
    return stk.empty();
}

int main() {
    // std::string str = "{[]}";
    // std::string str = "()";
    std::string str = "()[]{}";
    // std::string str = "(]";
    // std::string str = "([)]";
    bool result = isValid(str);
    std::cout << "Is parentheses valid? " << std::boolalpha << result << std::endl;
    return 0;
}